packages/clarity-js/src/data/extract.ts

```diff
@@ -0,0 +1,127 @@
+import encode from "./encode";
+import * as dom from "@src/layout/dom";
+import * as internal from "@src/diagnostic/internal";
+import { Code, Constant, Severity } from "@clarity-types/data";
+
+export let data: {[key: number]: string } = {};
+export let updateKeys: (number | string)[] = [];
+
+let variables : { [key: number]: string } = {};
+let selectors : { [key: number]: string } = {};
+
+export function start(): void {
+    let e = config.extract;
+    for (let i = 0; i < e.length; i+=3) {
+        let source = e[i] as Source;
+        let key = e[i+1] as number;
+        switch (source) {
+            case Source.Javascript: 
+                let variable = e[i+2] as string;
+                variables[key] = variable;
+                break;
+            case Source.Cookie: 
+                break;
+            case Source.Text:
+                let match = e[i+2] as string;
+                selectors[key] = match;
+                break;
+            case Source.Fragment:
+                let fragments =  e[i+2] as string[];
+                dom.setFragments(fragments);
+                break;
+        }
+    }
+}
+
+export function compute(): void {
+    try {
+        for (let v in variables) {
+            let value = evaluate(variables[v]);
+            if (value) { update(v, value); }
+        }
+
+        for (let s in selectors) {
+            let node = document.querySelector(selectors[s] as string) as HTMLElement;
+            if (node) { update(s, node.innerText); }
+        }
+                
+        let fragmentIds = dom.getFragments();
+        for (let hash in fragmentIds) {
+            update(hash, fragmentIds[hash], true);
+        }
+    }
+    catch (e) { internal.log(Code.Selector, Severity.Warning, e ? e.name : null); }
+
+    encode(Event.Extract);
+}
+
+export function reset(): void {
+    updateKeys = [];
+}
+
+function update(key: number | string, value: string | number, force: boolean = false): void {
+    if (!(key in data) || (key in data && data[key] !== value) || force ) {
+        data[key] = value;
+        updateKeys.push(key);
+    }
+}
+
+export function stop(): void {
+    data = {};
+    updateKeys = [];
+    variables = {};
+    selectors = {};
+}
+
+// The function below takes in a variable name in following format: "a.b.c" and safely evaluates its value in javascript context
+// For instance, for a.b.c, it will first check window["a"]. If it exists, it will recursively look at: window["a"]["b"] and finally,
+// return the value for window["a"]["b"]["c"].
+function evaluate(variable: string, type: string = null, base: Object = window): any {
+    let parts = variable.split(Constant.Dot);
+    let syntax = parts.shift();
+    let [first, array, condition] = parse(syntax);
+    let output;
+    if (base && base[first]) {
+        if (array === -1 && match(base[first], condition)) {
+            output = parts.length > 0 ? evaluate(parts.join(Constant.Dot), type, base[first]) : base[first]
```

Comment:
This seems a bit inefficient as it reparses, etc. each iteration through the expression. Can's you just iterate over the parts and update the base?